File: src/main/clojure/comzotohcljc/wflow/switch.clj
;;
;; COPYRIGHT (C) 2013 CHERIMOIA LLC. ALL RIGHTS RESERVED.
;;
;; THIS IS FREE SOFTWARE; YOU CAN REDISTRIBUTE IT AND/OR
;; MODIFY IT UNDER THE TERMS OF THE APACHE LICENSE
;; VERSION 2.0 (THE "LICENSE").
;;
;; THIS LIBRARY IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL
;; BUT WITHOUT ANY WARRANTY; WITHOUT EVEN THE IMPLIED WARRANTY OF
;; MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
;;
;; SEE THE LICENSE FOR THE SPECIFIC LANGUAGE GOVERNING PERMISSIONS
;; AND LIMITATIONS UNDER THE LICENSE.
;;
;; You should have received a copy of the Apache License
;; along with this distribution; if not you may obtain a copy of the
;; License at
;; http://www.apache.org/licenses/LICENSE-2.0
;;

(ns ^{ :doc ""
       :author "kenl" }
  comzotohcljc.wflow.switch )


(use '[clojure.tools.logging :only (info warn error debug)])
(import '(com.zotoh.hohenheim.core Job))

(require '[comzotohcljc.util.seqnumgen :as SN])
(require '[comzotohcljc.util.coreutils :as CU])

(use '[comzotohcljc.wflow.core])

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defprotocol SwitchPoint)
(defprotocol Switch)

(defprotocol SwitchChoiceExpr
  (getChoice [_ job] ))

(defn make-switch [choiceExpr]
  (let [ a (make-activity Switch) ]
    (.setf a :test choiceExpr)
    (.setf a :default nil)
    (.setf a :choices {} )
    a))

(defmethod ac-reify :Switch [ac cur]
  (ac-spawnpoint ac cur SwitchPoint))

(defmethod ac-realize! :Switch [ac cur]
  (let [ cs (.getf ac :choices)
         df (.getf cur :default)
         np (.getf cur :next)
         t  (reduce (fn [sum en]
                      (assoc sum (first en) (ac-reify (last en) np)) )
                    {} (seq cs)) ]
    (.setf cur :choices t)
    (when-not (nil? df)
      (.setf cur :default (ac-reify df np)))
    (.setf cur :test (.getf ac :test))
    cur))

(defmethod fw-evaluate! :SwitchPoint [fw job]
  (let [ cs (.getf fw :choices)
         df (.getf fw :default)
         c (fw-popattmt! fw)
         e (.getf fw :test) ]
    (let [ h (get cs (.getChoice e job))
           x (if (nil? h) df h) ]
      (when-not (nil? x) (fw-setattmt! x c))
      (fw-realize! fw)
      x)))


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;




(def ^:private switch-eof nil)

File: src/main/scala/com/zotoh/wflow/Switch.scala
/*??
*
* Copyright (c) 2013 Cherimoia, LLC. All rights reserved.
*
* This library is distributed in the hope that it will be useful
* but without any warranty; without even the implied warranty of
* merchantability or fitness for a particular purpose.
*
* The use and distribution terms for this software are covered by the
* Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
* which can be found in the file epl-v10.html at the root of this distribution.
*
* By using this software in any fashion, you are agreeing to be bound by
* the terms of this license.
* You must not remove this notice, or any other, from this software.
*
 ??*/


package com.zotoh.wflow

import scala.collection.JavaConversions._
import scala.collection.mutable


/**
 * @author kenl
 *
 */
class Switch(private var _expr:SwitchChoiceExpr ) extends Activity {

  private val _choices= mutable.HashMap[Any,Activity]()
  private var _def:Activity =null

  def withChoice(matcher:Any, body:Activity ): this.type = {
    _choices.put(matcher, body)
    this
  }

  def withDef(a:Activity ): this.type = {
    _def=a
    this
  }

  def reifyPoint(cur:FlowPoint ) = new SwitchPoint(cur, this)

  def realize(fp:FlowPoint ) {
    val t= mutable.HashMap[Any,FlowPoint]()
    val p=fp.asInstanceOf[SwitchPoint]
    val nxt= p.nextPoint()

    _choices.foreach { (en) =>
      t.put(en._1,
              en._2.reify(nxt))
    }

    p.withChoices(t.toMap)
    if (_def != null) {
      p.withDef( _def.reify(nxt))
    }
    p.withExpr(_expr)
  }








}
